### ISSUE TITLE
Unify link hints (a big idea, but a good one?)

### ISSUE THREAD

**Issue by smblott-github** [philc/vimium#2013]
With two hints modes, we get issues like #2011, or one somewhere where the user could choose the hints mode base on the URL.  Sometimes one mode works best, sometimes the other.  However:

- Almost none of the alphabet hints we generate are words (or possibly even prefixes of words, or parts of words).
- So, the text of an alphabet hint is unlikely to match the text of a link.
- And, text typed to match a link is very unlikely to match an alphabet hint.

Here's the idea:

- We have just one hints mode: usually with `asdfghjkl` as the hint keys.
- We generate hint strings as is done with alphabet hints.
- When the user types `keyChar`, we keep each hint where `keychar` is the next unmatched hint character, **or** where the typed text matches the link text.
- Whenever `keyChar` does not continue a hint, we re-label the hints (a la filtered hints).

If the user just types hint characters, this would (probably) behave almost exactly like alphabet hints.  And if they type link text, then almost exactly like filtered hints.  So, the best of both worlds.

The problem case is a hint like `a`, which matches both a hint and the text of other links.  If we use a scoring scheme similar to filtered hints, but score hint matches higher, then the user would have to hit `Enter` in this case.

Edit: Forgot to mention...  The link-hints code is so tangled that it's difficult to implement even a PoC of this.  So I'm floating the idea, before trying an implementation.

**Comment by mrmr1993** [comment]
This seems like a very complicated way to implement filtering. Can we not just expose a command in link hints mode that
* shows the HUD in 'find mode'
* live filters the links based on the HUD text
* if only 1 link remains, `<enter>` from the HUD also clicks the link.

Benefits:
* `f/save<enter>` is stable (assuming `/` is the default filter binding)
  - in your suggestion, `sa` is sometimes exactly selecting a link hint, which will cause this to select the wrong link seemingly randomly
  - the user can re-use a previous search, if we use `FindModeHistory` or similar
* internationalised filters (because HUD)
* I can still use `f` to exit link hint mode once it isn't in my link hint keys. (This breaking is a massive PITA for me)

Issues:
* Requires 2 extra keystrokes (`/` and `<enter>` in my mental model).
* HUD may cover a few hints.

**Comment by smblott-github** [comment]
I tried this out ([branch](https://github.com/smblott-github/vimium/tree/single-link-hints-mode), [commit](91ddcfdbe18db9b6d392d3cca73b98e26e6e59d7)).

It works so-so.  For typing links' texts, it's good.  For typing hints, not so good.  The problem is that a proper suffix of the fully-selected hint tends also to be a prefix of another hint.  So the hint isn't activated. Oh well.

I did learn a couple of things, though:

- Remarkably, 91ddcfdbe18db9b6d392d3cca73b98e26e6e59d7 pretty much implements all of filtered hints but *without* the filtered hints class.  So it would be pretty easy to collapse the two implementations.

- The technique used in 91ddcfdbe18db9b6d392d3cca73b98e26e6e59d7 for matching links ([here](https://github.com/smblott-github/vimium/commit/91ddcfdbe18db9b6d392d3cca73b98e26e6e59d7#diff-e9abcb9ebcdb5af8e9f33651364673a1R505)) might allow us to filter link texts which include the hint characters.  E.g., match `7eleven` by typing `7e`.
